model.py

```diff
@@ -476,316 +450,153 @@ def service_area_name(self):
+        #         at the poles.
+        radius_in_meters   = max_radius * 1000.0                # noqa: E221
+        northeast_angle    = func.radians(45.0)                 # noqa: E221  # Radian angle from north (0), clockwise
+        target_geometry    = cast(target.ewkt, Geometry)        # noqa: E221  # ST_Distance wants a geometry
+        target_geography   = cast(target.ewkt, Geography)       # noqa: E221  # ST_Project wants a geography
+        northeast_ref_pt   = func.ST_Project(target_geography, radius_in_meters, northeast_angle)  # noqa: E221
+        ne_ref_pt_geom     = cast(northeast_ref_pt, Geometry)   # noqa: E221
+
+        # This MUST have two Geometry parameters. If it uses Geography the distance returned
+        # will be in meters, NOT radians.
+        local_dist_in_rads = func.ST_Distance(target_geometry, ne_ref_pt_geom)
+
+        # Step 2: Use the localized value to set up a geometry-based radius function
+        #         that's locally accurate (for rough values of 'accurate'). This will
+        #         let us limit our search to Places no further than max_radius km from
+        #         our target location.
+        within_max_radius = func.ST_DWithin(target_geometry, Place.geometry, local_dist_in_rads)
+
+        # Step 3: Set up a function to find the minimum distance between the
+        #         target (a point), and an edge of the Place objects we want
+        #         to search, which are typically polygons. We won't restrict
+        #         the query on this, but we do want to return it as a column,
+        #         and use it to order results.
+        meters_to_near_edge = func.min(func.ST_DistanceSphere(target_geometry, Place.geometry))
+
+        # Step 4: Establish the query object, and restrict it to the right feed.
+        #         We only want to search against the geometry of Place objects which
+        #         are used as the service area of one or more Libraries, to avoid
+        #         searching against all Places loaded in the database, so we join
+        #         Library -> ServiceArea -> Place.
+        query_obj = db_session.query(Library).join(Library.service_areas).join(ServiceArea.place)
+        query_obj = query_obj.filter(cls._feed_restriction(production))
+
+        # Step 5: Use our radius function to add a WHERE clause, so that we return
+        #         only Library objects within the max_radius in km.
+        query_obj = query_obj.filter(within_max_radius)
+
+        # Step 6: Add the distance from our target point to the near edge as a
+        #         column in the output, group our results by Library, and then
+        #         use the point-to-polygon-edge distance to order our result set.
+        query_obj = query_obj.add_columns(meters_to_near_edge)
+        query_obj = query_obj.group_by(Library.id)
+        query_obj = query_obj.order_by(meters_to_near_edge.asc())
```

Comment:
Just a note for later unless you fixed this, but this GROUP BY might explain why libraries sometimes show up twice in the results -- they have two different service areas that both match the criteria.

Reply:
Oh, interesting. I'll have to figure out how to consolidate that and test for it. Good catch.